File: iOS/MatrixMath/MatrixMath/Matrix.swift
//
//  Matrix.swift
//  MatrixMath
//
//  Created by Ivan Magda on 02.04.16.
//  Copyright © 2016 Ivan Magda. All rights reserved.
//

import Foundation

/// Defining the matrix as array of arrays with specific type of the elements.
typealias MatrixDataType = Array<Array<Double>>

//------------------------------------
// MARK: - Matrix
//------------------------------------

struct Matrix {
    
    //------------------------------------
    // MARK: Properties
    //------------------------------------
    
    let data: MatrixDataType
    let dimension: MatrixDimension
    
    //------------------------------------
    // MARK: Initializers
    //------------------------------------
    
    init(data: MatrixDataType) {
        self.data = data
        
        let columns = data.count
        let rows = data[0].count
        self.dimension = MatrixDimension(columns: columns, rows: rows)
    }
    
    init?(data: Array<Double>, dimension: MatrixDimension) {
        guard data.count == dimension.count()
            && data.count > 0
            && dimension.nonZero() else {
            return nil
        }
        
        var array = MatrixDataType()
        for columnIdx in 0..<dimension.columns {
            var elemIdx = columnIdx * dimension.rows
            var elements = [Double]()
            for _ in 0..<dimension.rows {
                elements.append(data[elemIdx])
                elemIdx += 1
            }
            array.append(elements)
        }
        
        self.data = array
        self.dimension = dimension
    }
    
    //------------------------------------
    // MARK: Behavior
    //------------------------------------
    
    func getLinearArray() -> [Double] {
        var result = [Double]()
        for anArray in data {
            for element in anArray {
                result.append(element)
            }
        }
        
        return result
    }
    
}

File: Android/MatrixMath/app/src/main/java/com/ivanmagda/matrixmath/model/Matrix.java
package com.ivanmagda.matrixmath.model;

import java.util.ArrayList;
import java.util.List;

public class Matrix {

    private List<List<Double>> elements;
    private MatrixDimension dimension;

    public Matrix(List<List<Double>> elements) {
        this.elements = elements;

        int columns = elements.size();
        int rows = elements.get(0).size();
        this.dimension = new MatrixDimension(columns, rows);
    }

    public Matrix(List<Double> data, MatrixDimension dimension) {
        if (dimension.getCount() != data.size()) throw new AssertionError();

        List<List<Double>> res = new ArrayList<>(dimension.getRows());
        for (int columnIdx = 0; columnIdx < dimension.getColumns(); columnIdx++) {
            int elemIdx = columnIdx * dimension.getRows();
            List<Double> elements = new ArrayList<>(dimension.getColumns());
            for (int i = 0; i < dimension.getRows(); i++) {
                elements.add(data.get(elemIdx));
                elemIdx++;
            }
            res.add(elements);
        }

        this.elements = res;
        this.dimension = dimension;
    }

    public List<List<Double>> getElements() {
        return elements;
    }

    public MatrixDimension getDimension() {
        return dimension;
    }

    public List<Double> getOneDimensionArrayOfElements() {
        List<Double> result = new ArrayList<>(dimension.getCount());
        for (List<Double> anArray : elements)
            for (Double element : anArray)
                result.add(element);
        return result;
    }

}
